File: analytics/src/main/java/com/segment/analytics/AnalyticsActivityLifecycleCallbacks.java
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Segment.io, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.segment.analytics;

import android.app.Activity;
import android.app.Application;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import com.segment.analytics.internal.Utils;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

class AnalyticsActivityLifecycleCallbacks
        implements Application.ActivityLifecycleCallbacks, DefaultLifecycleObserver {
    private Analytics analytics;
    private ExecutorService analyticsExecutor;
    private Boolean shouldTrackApplicationLifecycleEvents;
    private Boolean trackDeepLinks;
    private Boolean shouldRecordScreenViews;
    private PackageInfo packageInfo;

    private AtomicBoolean trackedApplicationLifecycleEvents;
    private AtomicInteger numberOfActivities;
    private AtomicBoolean firstLaunch;

    private AtomicBoolean isChangingActivityConfigurations;
    private Boolean useNewLifecycleMethods;

    // This is just a stub LifecycleOwner which is used when we need to call some lifecycle
    // methods without going through the actual lifecycle callbacks
    private static LifecycleOwner stubOwner =
            new LifecycleOwner() {
                Lifecycle stubLifecycle =
                        new Lifecycle() {
                            @Override
                            public void addObserver(@NonNull LifecycleObserver observer) {
                                // NO-OP
                            }

                            @Override
                            public void removeObserver(@NonNull LifecycleObserver observer) {
                                // NO-OP
                            }

                            @NonNull
                            @Override
                            public Lifecycle.State getCurrentState() {
                                return State.DESTROYED;
                            }
                        };

                @NonNull
                @Override
                public Lifecycle getLifecycle() {
                    return stubLifecycle;
                }
            };

    private AnalyticsActivityLifecycleCallbacks(
            Analytics analytics,
            ExecutorService analyticsExecutor,
            Boolean shouldTrackApplicationLifecycleEvents,
            Boolean trackDeepLinks,
            Boolean shouldRecordScreenViews,
            PackageInfo packageInfo,
            Boolean useNewLifecycleMethods) {
        this.trackedApplicationLifecycleEvents = new AtomicBoolean(false);
        this.numberOfActivities = new AtomicInteger(1);
        this.firstLaunch = new AtomicBoolean(false);
        this.analytics = analytics;
        this.analyticsExecutor = analyticsExecutor;
        this.shouldTrackApplicationLifecycleEvents = shouldTrackApplicationLifecycleEvents;
        this.trackDeepLinks = trackDeepLinks;
        this.shouldRecordScreenViews = shouldRecordScreenViews;
        this.packageInfo = packageInfo;
        this.useNewLifecycleMethods = useNewLifecycleMethods;
        this.isChangingActivityConfigurations = new AtomicBoolean(false);
    }

    @Override
    public void onStop(@NonNull LifecycleOwner owner) {
        // App in background
        if (shouldTrackApplicationLifecycleEvents
                && numberOfActivities.decrementAndGet() == 0
                && !isChangingActivityConfigurations.get()) {
            analytics.track("Application Backgrounded");
        }
    }

    @Override
    public void onStart(@NonNull LifecycleOwner owner) {
        // App in foreground
        if (shouldTrackApplicationLifecycleEvents
                && numberOfActivities.incrementAndGet() == 1
                && !isChangingActivityConfigurations.get()) {
            Properties properties = new Properties();
            if (firstLaunch.get()) {
                properties
                        .putValue("version", packageInfo.versionName)
                        .putValue("build", String.valueOf(packageInfo.versionCode));
            }
            properties.putValue("from_background", !firstLaunch.getAndSet(false));
            analytics.track("Application Opened", properties);
        }
    }

    @Override
    public void onCreate(@NonNull LifecycleOwner owner) {
        // App created
        if (!trackedApplicationLifecycleEvents.getAndSet(true)
                && shouldTrackApplicationLifecycleEvents) {
            numberOfActivities.set(0);
            firstLaunch.set(true);
            analytics.trackApplicationLifecycleEvents();
        }
    }

    @Override
    public void onResume(@NonNull LifecycleOwner owner) {}

    @Override
    public void onPause(@NonNull LifecycleOwner owner) {}

    @Override
    public void onDestroy(@NonNull LifecycleOwner owner) {}

    @Override
    public void onActivityCreated(Activity activity, Bundle bundle) {
        analytics.runOnMainThread(IntegrationOperation.onActivityCreated(activity, bundle));

        if (!useNewLifecycleMethods) {
            onCreate(stubOwner);
        }

        if (trackDeepLinks) {
            trackDeepLink(activity);
        }
    }

    private void trackDeepLink(Activity activity) {
        Intent intent = activity.getIntent();
        if (intent == null || intent.getData() == null) {
            return;
        }

        Properties properties = new Properties();

        Uri referrer = Utils.getReferrer(activity);
        if (referrer != null) {
            properties.putReferrer(referrer.toString());
        }

        Uri uri = intent.getData();
        try {
            for (String parameter : uri.getQueryParameterNames()) {
                String value = uri.getQueryParameter(parameter);
                if (value != null && !value.trim().isEmpty()) {
                    properties.put(parameter, value);
                }
            }
        } catch (Exception e) {
            analytics
                    .logger("LifecycleCallbacks")
                    .error(e, "failed to get uri params for %s", uri.toString());
        }

        properties.put("url", uri.toString());
        analytics.track("Deep Link Opened", properties);
    }

    @Override
    public void onActivityStarted(Activity activity) {
        if (shouldRecordScreenViews) {
            analytics.recordScreenViews(activity);
        }
        analytics.runOnMainThread(IntegrationOperation.onActivityStarted(activity));
    }

    @Override
    public void onActivityResumed(Activity activity) {
        analytics.runOnMainThread(IntegrationOperation.onActivityResumed(activity));
        if (!useNewLifecycleMethods) {
            onStart(stubOwner);
        }
    }

    @Override
    public void onActivityPaused(Activity activity) {
        analytics.runOnMainThread(IntegrationOperation.onActivityPaused(activity));
        if (!useNewLifecycleMethods) {
            onPause(stubOwner);
        }
    }

    @Override
    public void onActivityStopped(Activity activity) {
        analytics.runOnMainThread(IntegrationOperation.onActivityStopped(activity));
        if (!useNewLifecycleMethods) {
            onStop(stubOwner);
        }
    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle bundle) {
        analytics.runOnMainThread(
                IntegrationOperation.onActivitySaveInstanceState(activity, bundle));
    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        analytics.runOnMainThread(IntegrationOperation.onActivityDestroyed(activity));
        if (!useNewLifecycleMethods) {
            onDestroy(stubOwner);
        }
    }

    public static class Builder {
        private Analytics analytics;
        private ExecutorService analyticsExecutor;
        private Boolean shouldTrackApplicationLifecycleEvents;
        private Boolean trackDeepLinks;
        private Boolean shouldRecordScreenViews;
        private PackageInfo packageInfo;
        private Boolean useNewLifecycleMethods;

        public Builder() {}

        public Builder analytics(Analytics analytics) {
            this.analytics = analytics;
            return this;
        }

        Builder analyticsExecutor(ExecutorService analyticsExecutor) {
            this.analyticsExecutor = analyticsExecutor;
            return this;
        }

        Builder shouldTrackApplicationLifecycleEvents(
                Boolean shouldTrackApplicationLifecycleEvents) {
            this.shouldTrackApplicationLifecycleEvents = shouldTrackApplicationLifecycleEvents;
            return this;
        }

        Builder trackDeepLinks(Boolean trackDeepLinks) {
            this.trackDeepLinks = trackDeepLinks;
            return this;
        }

        Builder shouldRecordScreenViews(Boolean shouldRecordScreenViews) {
            this.shouldRecordScreenViews = shouldRecordScreenViews;
            return this;
        }

        Builder packageInfo(PackageInfo packageInfo) {
            this.packageInfo = packageInfo;
            return this;
        }

        Builder useNewLifecycleMethods(boolean useNewLifecycleMethods) {
            this.useNewLifecycleMethods = useNewLifecycleMethods;
            return this;
        }

        public AnalyticsActivityLifecycleCallbacks build() {
            return new AnalyticsActivityLifecycleCallbacks(
                    analytics,
                    analyticsExecutor,
                    shouldTrackApplicationLifecycleEvents,
                    trackDeepLinks,
                    shouldRecordScreenViews,
                    packageInfo,
                    useNewLifecycleMethods);
        }
    }
}

File: analytics/src/main/java/com/segment/analytics/AnalyticsActivityLifecycleCallbacks.kt
package com.segment.analytics

import android.app.Activity
import android.app.Application
import android.content.SharedPreferences
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import com.segment.analytics.integrations.Logger
import com.segment.analytics.internal.Private
import com.segment.analytics.platform.AndroidLifecycle
import com.segment.analytics.platform.Extension
import java.util.concurrent.ExecutorService
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

internal class AnalyticsActivityLifecycleCallbacks(
        private val analyticsObj: Analytics,
        private val analyticsExecutor: ExecutorService,
        private val shouldTrackApplicationLifecycleEvents: Boolean,
        private val trackDeepLinks: Boolean,
        private val shouldRecordScreenViews: Boolean,
        private val packageInfo: PackageInfo,
        private val useNewLifecycleMethods: Boolean,
        private val sharedPreferences: SharedPreferences,
        private val logger: Logger
) : Application.ActivityLifecycleCallbacks, DefaultLifecycleObserver, Extension {

    override val type = Extension.Type.Before
    override val name = "AnalyticsActivityLifecycleCallbacksExtension"
    override var analytics: Analytics? = analyticsObj
    private val trackedApplicationLifecycleEvents = AtomicBoolean(false)
    private val numberOfActivities = AtomicInteger(1)
    private val firstLaunch = AtomicBoolean(false)
    private val isChangingActivityConfigurations = AtomicBoolean(false)

//    private val application: Application
//    private val lifecycle: Lifecycle
//
//    init {
//        setupListeners()
//    }
//
//    private fun setupListeners() {
//        application.registerActivityLifecycleCallbacks(this)
//        if (useNewLifecycleMethods) {
//            lifecycle.addObserver(this)
//        }
//    }

    private fun runOnAnalyticsThread(block: () -> Unit) {
        if (!analyticsObj.isShutdown) {
            analyticsExecutor.submit(block)
        }
    }

    /* OLD LIFECYCLE HOOKS */
    override fun onActivityCreated(activity: Activity?, bundle: Bundle?) {
        runOnAnalyticsThread {
            analyticsObj.timeline?.applyClosure { extension: Extension? ->
                if (extension is AndroidLifecycle) {
                    (extension as AndroidLifecycle).onActivityCreated(activity, bundle)
                }
            }
        }
        if (!useNewLifecycleMethods) {
            onCreate(stubOwner)
        }
        if (trackDeepLinks) {
            trackDeepLink(activity)
        }
    }

    override fun onActivityStarted(activity: Activity?) {
        if (shouldRecordScreenViews) {
            recordScreenViews(activity)
        }
        runOnAnalyticsThread {
            analyticsObj.timeline.applyClosure { extension: Extension? ->
                if (extension is AndroidLifecycle) {
                    (extension as AndroidLifecycle).onActivityStarted(activity)
                }
            }
        }
    }

    override fun onActivityResumed(activity: Activity?) {
        runOnAnalyticsThread {
            analyticsObj.timeline.applyClosure { extension: Extension? ->
                if (extension is AndroidLifecycle) {
                    (extension as AndroidLifecycle).onActivityResumed(activity)
                }
            }
        }
        if (!useNewLifecycleMethods) {
            onStart(stubOwner)
        }
    }

    override fun onActivityPaused(activity: Activity?) {
        runOnAnalyticsThread {
            analyticsObj.timeline.applyClosure { extension: Extension? ->
                if (extension is AndroidLifecycle) {
                    (extension as AndroidLifecycle).onActivityPaused(activity)
                }
            }
        }
        if (!useNewLifecycleMethods) {
            onPause(stubOwner)
        }
    }

    override fun onActivityStopped(activity: Activity?) {
        runOnAnalyticsThread {
            analyticsObj.timeline.applyClosure { extension: Extension? ->
                if (extension is AndroidLifecycle) {
                    (extension as AndroidLifecycle).onActivityStopped(activity)
                }
            }
        }
        if (!useNewLifecycleMethods) {
            onStop(stubOwner)
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity?, bundle: Bundle?) {
        runOnAnalyticsThread {
            analyticsObj.timeline.applyClosure { extension: Extension? ->
                if (extension is AndroidLifecycle) {
                    (extension as AndroidLifecycle).onActivitySaveInstanceState(activity, bundle)
                }
            }
        }
    }

    override fun onActivityDestroyed(activity: Activity?) {
        runOnAnalyticsThread {
            analyticsObj.timeline.applyClosure { extension: Extension? ->
                if (extension is AndroidLifecycle) {
                    (extension as AndroidLifecycle).onActivityDestroyed(activity)
                }
            }
        }
        if (!useNewLifecycleMethods) {
            onDestroy(stubOwner)
        }
    }

    /* NEW LIFECYCLE HOOKS (These get called alongside the old ones) */

    override fun onStop(owner: LifecycleOwner) {
        // App in background
        if (shouldTrackApplicationLifecycleEvents
                && numberOfActivities.decrementAndGet() == 0 && !isChangingActivityConfigurations.get()) {
            analyticsObj.track("Application Backgrounded")
        }
    }

    override fun onStart(owner: LifecycleOwner) {
        // App in foreground
        if (shouldTrackApplicationLifecycleEvents
                && numberOfActivities.incrementAndGet() == 1 && !isChangingActivityConfigurations.get()) {
            val properties = Properties()
            if (firstLaunch.get()) {
                properties
                        .putValue("version", packageInfo.versionName)
                        .putValue("build", packageInfo.versionCode.toString())
            }
            properties.putValue("from_background", !firstLaunch.getAndSet(false))
            analyticsObj.track("Application Opened", properties)
        }
    }

    override fun onCreate(owner: LifecycleOwner) {
        // App created
        if (!trackedApplicationLifecycleEvents.getAndSet(true)
                && shouldTrackApplicationLifecycleEvents) {
            numberOfActivities.set(0)
            firstLaunch.set(true)
            trackApplicationLifecycleEvents()
        }
    }

    override fun onResume(owner: LifecycleOwner) {}
    override fun onPause(owner: LifecycleOwner) {}
    override fun onDestroy(owner: LifecycleOwner) {}

    private fun trackDeepLink(activity: Activity?) {
        val intent = activity?.intent
        if (intent == null || intent.data == null) {
            return
        }
        val properties = Properties()
        val uri = intent.data
        uri?.let {
            for (parameter in uri.queryParameterNames) {
                val value = uri.getQueryParameter(parameter)
                if (value != null && value.trim().isNotEmpty()) {
                    properties[parameter] = value
                }
            }
            properties["url"] = uri.toString()
        }
        analyticsObj.track("Deep Link Opened", properties)
    }

    @Private
    fun trackApplicationLifecycleEvents() {
        // Get the current version.
        val packageInfo = packageInfo
        val currentVersion = packageInfo.versionName
        val currentBuild = packageInfo.versionCode

        // Get the previous recorded version.
        val sharedPreferences = sharedPreferences
        val previousVersion = sharedPreferences.getString(VERSION_KEY, null)
        val previousBuild = sharedPreferences.getInt(BUILD_KEY, -1)

        // Check and track Application Installed or Application Updated.
        if (previousBuild == -1) {
            analyticsObj.track(
                    "Application Installed",
                    Properties() //
                            .putValue(VERSION_KEY, currentVersion)
                            .putValue(BUILD_KEY, currentBuild.toString()))
        } else if (currentBuild != previousBuild) {
            analyticsObj.track(
                    "Application Updated",
                    Properties() //
                            .putValue(VERSION_KEY, currentVersion)
                            .putValue(BUILD_KEY, currentBuild.toString())
                            .putValue("previous_$VERSION_KEY", previousVersion)
                            .putValue("previous_$BUILD_KEY", previousBuild.toString()))
        }

        // Update the recorded version.
        val editor = sharedPreferences.edit()
        editor.putString(VERSION_KEY, currentVersion)
        editor.putInt(BUILD_KEY, currentBuild)
        editor.apply()
    }

    @Private
    fun recordScreenViews(activity: Activity?) {
        val packageManager = activity?.packageManager
        try {
            val info = packageManager?.getActivityInfo(
                    activity.componentName, PackageManager.GET_META_DATA)
            val activityLabel = info?.loadLabel(packageManager)
            analyticsObj.screen(activityLabel.toString())
        } catch (e: PackageManager.NameNotFoundException) {
            throw AssertionError("Activity Not Found: $e")
        } catch (e: Exception) {
            logger.error(e, "Unable to track screen view for %s", activity.toString())
        }
    }

    companion object {
        private const val VERSION_KEY = "version"
        private const val BUILD_KEY = "build"

        // This is just a stub LifecycleOwner which is used when we need to call some lifecycle
        // methods without going through the actual lifecycle callbacks
        private val stubOwner: LifecycleOwner = object : LifecycleOwner {
            var stubLifecycle: Lifecycle = object : Lifecycle() {
                override fun addObserver(observer: LifecycleObserver) {
                    // NO-OP
                }

                override fun removeObserver(observer: LifecycleObserver) {
                    // NO-OP
                }

                override fun getCurrentState(): State {
                    return State.DESTROYED
                }
            }

            override fun getLifecycle(): Lifecycle {
                return stubLifecycle
            }
        }
    }

}
